src/test/ui/mir/issue-105809.rs

```diff
@@ -0,0 +1,36 @@
+// Non-regression test ICE from issue #105809 and duplicates.
+
+// build-pass: the ICE is during codegen
+// compile-flags: --edition 2018
```

Comment:
Can you add `-Zmir-opt-level=1` here? It was reported in one of the issues that this only ICE'd with that option.

Reply:
I can, but IIRC 1 is the default, and this test does ICE as is on master.

So I take it we'll land the revert and won't have time to do a fix ?

Reply:
I'm looking into it right now, but am not sure whether I can fix it in the next couple of hours. If I don't open a PR in the next two hours we can revert it.

Reply:
Don't worry about it too much, it's perfectly fine, we can land the revert to buy us time to look into it without pressure.

I've now added `-Zmir-opt-level=1` to the test.

Reply:
Getting somewhat closer to figuring out what is going wrong. Can you give me 30 more minutes?

Reply:
Ah but of course! #105881 and #105886 don't have MCVEs yet so hopefully these 3 issues all have the same underlying issue.

I've marked this PR as ready, and if you want it to land, either you or @jackh726 can tell bors that whenever you need. The time frame jack refers to here https://github.com/rust-lang/rust/pull/105905#issuecomment-1357914851 would be 2AM my time.

Reply:
We add an additional `OpaqueCast` [here](https://github.com/rust-lang/rust/blob/7ab803891dd95a04f3b2143e40b26f03b608b4bc/compiler/rustc_mir_build/src/build/matches/util.rs#L109-L115) because `ty != pattern.ty`. This then later ICEs because we cannot get the layout for that type, I don't understand why getting the layout fails though. Will try to work more on this in the next few days.

Thanks for opening the revert PR @lqd.